src/Transform/ONNX/ONNXPreKrnlVerifyPass.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------- ONNXPreKrnlVerifyPass.cpp - Verification -------------------===//
+//
+// Copyright 2019-2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file implements a Function level pass that verifies whether
+// ONNX ops in the function are ready for lowering to Krnl.
+//
+//===----------------------------------------------------------------------===//
+
+#include <set>
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Interfaces/CallInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "src/Dialect/Krnl/KrnlOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Interface/ShapeInferenceOpInterface.hpp"
+#include "src/Pass/Passes.hpp"
+#include "src/Support/OMOptions.hpp"
+
+using namespace mlir;
+
+namespace {
+
+/*!
+ * This pass verify whether each ONNX op is ready for Krnl
+ * The current condition is that all input tensors have to be ranked
+ */
+
+class ONNXPreKrnlVerifyPass
+    : public mlir::PassWrapper<ONNXPreKrnlVerifyPass, FunctionPass> {
+
+public:
+  void runOnFunction() override {
+    auto function = getFunction();
+    auto &funcBody = function.getBody();
+
+    // Iterate on the operations
+    for (Operation &op : funcBody.getOps()) {
+      if (isa<mlir::ONNXOpsDialect>(op.getDialect())) {
+        if (failed(verifyRanked(op)))
+          signalPassFailure();
+      }
+    }
+  }
+
+private:
+  static LogicalResult verifyRanked(Operation &op) {
+    for (auto ty : op.getOperandTypes()) {
+      if (!ty.isa<RankedTensorType>() && !ty.isa<NoneType>()) {
```

Comment:
Here is where all the previous errors are reduced to "ranked or not". But several errors where of a different nature.

Reply:
In shape inference, we did some semantic check with assumption of all tensors are statically shaped. This is approach may cause trouble. I only deleted the check for concatOp, which caused trouble for yolov4. I can not fix all them. To me it is not important to do the check and we can assume (or use onnx check model tool) the model is correct.

Reply:
Agreed, if other checks are more of the type "verify", it should be done in a verifier.